### DISCUSSION TITLE
HSM and Periodic control loops

### DISCUSSION THREAD

**Discussion by nens90** [igor-krechetov/hsmcpp#13] [Q&A]
Hi @igor-krechetov

Really cool project you have here 🤩

I'm interested in using your framework for a HSM which should manage different control strategies based on user input and a feedback loop.

The controller should run at a specific frequency every 1 sec.

Would you suggest to run the controller function as a callback in the HSM framework or run it in a separate task? 

I see that my question has some similarities to the discussion on Game rendering https://github.com/igor-krechetov/hsmcpp/discussions/6

I was wondering why you suggested to use the getActiveState() API. In your documentation you state that is more a debugging feature [Active States](https://hsmcpp.readthedocs.io/en/latest/features/substates/substates.html?highlight=Get%20active%20state#active-states)

Thank you in advance.

**Comment by igor-krechetov** [comment]
Hello @nens90 . Sorry for the late reply and thank you for the compliment. Glad to know that the library is useful for other developers. :)

As for your question, since your interval is quite high (1s) I'd recommend using HSM timer and callbacks. The reason I recommend to use a separate thread and getActiveStates API in a different discussion was because the requirements were different:
- low frequency (at least every 33ms for 30fps or even faster)
- rendering scene composition based on active states (triggering callbacks both for child and parent states at the same time would have required a workaround)

Using getActiveStates and a separate thread would have allowed to make sure that UI rendering does not depend on the main logic and can be executed in a predictable manner rather than relying on the amount of pending hsm events and any potentially slow callbacks.
